emscripten.py

```diff
@@ -498,13 +497,43 @@ def finalize_wasm(infile, outfile, memfile, js_syms):
   if settings.DEBUG_LEVEL >= 3:
     args.append('--dwarf')
 
-  metadata = get_metadata(infile, outfile, modify_wasm, args)
-  if modify_wasm:
-    building.save_intermediate(infile, 'post_finalize.wasm')
-  elif infile != outfile:
+  if infile != outfile:
     shutil.copy(infile, outfile)
+
+  if settings.GENERATE_SOURCE_MAP:
+    building.emit_wasm_source_map(infile, outfile + '.map', outfile)
+    building.save_intermediate(outfile + '.map', 'base_wasm.map')
+    base_url = settings.SOURCE_MAP_BASE + os.path.basename(outfile) + '.map'
+    if modify_wasm:
+      # If we are already modifying, just let Binaryen add the sourcemap URL
+      args += ['--output-source-map-url=' + base_url]
+    else:
+      # Otherwise use objcopy. This avoids re-encoding the file (thus
+      # preserving DWARF) and is faster.
+
+      # Create a file with the contents of the sourceMappingURL section
+      with shared.get_temp_files().get_file('.bin') as url_file:
+        utils.write_binary(url_file,
+                           leb128.u.encode(len(base_url)) + base_url.encode('utf-8'))
+        cmd = [shared.LLVM_OBJCOPY,
+               '--add-section',
+               'sourceMappingURL=' + url_file,
```

Comment:
It looks like this puts the location of a temporary file in this section? But I think we want it to point to the final location of the source map?

(In fact I'm confused as to why we want a temp file here?)

Reply:
The location of the file is not relevant.  Its just the `--add-section'` requires a file the contains the section contents.

Reply:
Thanks, I see, it just reads the contents... I misunderstood then.

Reply:
I was confused by this too at first. It turns out, `llvm-objcopy`'s `--add-section` option takes a file name whose contents will be put in the requested section. From [`llvm-objcopy`'s manual page](https://llvm.org/docs/CommandGuide/llvm-objcopy.html):
```
--add-section <section=file>
Add a section named <section> with the contents of <file> to the output. For ELF objects the section will be of type SHT_NOTE, if the name starts with “.note”. Otherwise, it will have type SHT_PROGBITS. Can be specified multiple times to add multiple sections.

For MachO objects, <section> must be formatted as <segment name>,<section name>.
```
So we have to create a temp file and write the URL of the map file into this temp file.

Reply:
Oops, I laced with Sam.